### DISCUSSION TITLE
tutorials/audio/sync_with_audio

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#217] [User-contributed notes]
# tutorials/audio/sync_with_audio

Introduction: In any application or game, sound and music playback will have a slight delay. For games, this delay is often so small that it is negligible. Sound effects will come out a few millise...

https://docs.godotengine.org/en/stable/tutorials/audio/sync_with_audio.html

**Comment by waso-mu** [comment]
**Quick word of warning from a rhythm game developer:**
AFAIK, you currently cannot accurately sync audio to gameplay, (only gameplay to audio as seen in this article). There's no way to schedule audio in advance, so you are limited to the user's framerate. For most games, not an issue. But if you are making music tools or a strict rhythm game, the delay is incredibly noticeable (esp on lower end devices) and can be unfair if not just offputting. You may be best off using a separate audio engine for your project if it requires accurate timed playback of rhythm.

**Reply by waso-mu** [reply]
I phrased the first sentence poorly, I mean you cannot accurately sync playback of audio to music, for example in the case of overlaying a hit-sound, metronome, or count-in.

**Reply by Calinou** [reply]
See proposal:

- https://github.com/godotengine/godot-proposals/issues/1151

**Reply by beatboxchad** [reply]
So I skimmed through all those PR's about the audio subsystem, starting with the ones linked. Really cool work. Valiant hacking, and I learned a lot. But I strongly suspect that a much more elegant solution will work for most all of us writing audio rhythm games:

For games where the audio is related to the physics, process the audio in [physics_process](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-private-method-physics-process), and drop whatever audio frames you don't render on time. It guarantees a flat sample rate and a timed call order, so scheduling events should be pretty easy too.

You don't have to do ALL of the audio in `physics_process`, just the events that are synced to gameplay and are, thus, part of the physics to process. `physics_process` is still sacred, but the argument is that in these types of games, some of the audio is part of the physics. `physics_process` provides all the guarantees needed. You'd probably wanna process those MIDI input events here too? For `AudioStreamGenerator`, you might set the playback buffer size to size of the `phsyics_process` window in terms of the audio rate (usually 44.1k). Maybe add a little padding to avoid dropped samples.

Make sense? It sure does to me. `Node.process` runs before video rendering and is tied to the game's frame rate at runtime, but has no other timing guarantees by design. Since there's no guarantee that a frame won't lag and cause your batch of samples to fluctuate in size, of course you'll have to fight to get audio logic defined there to sync with gameplay. Sure, put everything else in `process` and if the game visuals lag, fine, but the band played on. I'm gonna try this in my project and report back, but I wanted to discuss it here in case a more seasoned geek could point out a fatal flaw in my plan.

**Comment by waso-mu** [comment]
Here's a handy function code to convert seconds to crotchets (AKA beats/quarter notes):

```gdscript
# This dictionary stores each time the bpm changes in the song. It's in dictionary format since I wrote it to support custom levels stored in .json files. It does not support gradual BPM changes.
var bpm_changes:Array[Dictionary] = [ # Example value, sets the BPM to 120 at beat one and nothing else
    {
        "Crotchet":1.0, # The beat that the BPM changes
        "BPM":120,      # The BPM it changes to
    },
]

func seconds_to_crotchet(seconds:float) -> float:
	var remaining_seconds := seconds # I don't think this is necessary in gdscript, but just to be safe
	var last_bpm := 100.0 # arbitrary value in case no bpm is set on beat 1
	var last_crotchet := 1.0
	var total_crotchets := 1.0
	
	for change in bpm_changes:
		var full = remaining_seconds*last_bpm/60
		if full >= change["Crotchet"]:
			var diff = change["Crotchet"]-last_crotchet
			total_crotchets += diff
			remaining_seconds -= diff/last_bpm*60
			last_bpm = change["BPM"]
			last_crotchet = change["Crotchet"]
		else:
			break
	total_crotchets += remaining_seconds*last_bpm/60 + 1
	
	return total_crotchets
```
